### PR TITLE
update rule use-after-free, add rule function-use-after-free

### PR DESCRIPTION
**Pull request by infractal**
- update use-after-free for different pointer dereference patterns + handle re-malloc() case
- add rule function-use-after-free for use-after-free pointers passed to functions

### CODE REVIEW

File: c/lang/security/function-use-after-free.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+void other_func(char *ignored) {}
+void refreencode(char *ignored) {}
+
+int bad_code1() {
+    NAME *var;
+    char buf[10];
+    var = (NAME *)malloc(sizeof(struct name));
+    free(var);
+    // ruleid: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:18]:
```suggestion
    // ruleid: function-use-after-free
```

File: c/lang/security/function-use-after-free.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    strcpy(buf, (char*)var);
+    // ruleid: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:20]:
```suggestion
    // ruleid: function-use-after-free
```

File: c/lang/security/function-use-after-free.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    other_func((char*)(*var));
+    other_func((char*)var[0]);
+    // ruleid: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:23]:
```suggestion
    // ruleid: function-use-after-free
```

File: c/lang/security/function-use-after-free.c [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+int okay_code2() {
+    NAME *var;
+    var = (NAME *)malloc(sizeof(struct name));
+    free(var);
+    var = NULL;
+    var = (NAME *)malloc(sizeof(struct name));
+    // This will segmentation fault
+    // ok: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:63]:
```suggestion
    // ok: function-use-after-free
```

File: c/lang/security/function-use-after-free.c [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    return 0;
+}
+
+int okay_code1() {
+    NAME *var;
+    var = (NAME *)malloc(sizeof(struct name));
+    free(var);
+    var = NULL;
+    // This will segmentation fault
+    // ok: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:49]:
```suggestion
    // ok: function-use-after-free
```

**Comment by infractal** [comment]
Closing due to failed tests.

File: c/lang/security/function-use-after-free.c [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+int bad_code2() {
+    NAME *var;
+    char buf[10];
+    var = (NAME *)malloc(sizeof(struct name));
+    free(var);
+    // ruleid: use-after-free
+    strcpy(buf, (char*)*var);
+    // ruleid: use-after-free
+    other_func((char*)var);
+    // ruleid: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:38]:
```suggestion
    // ruleid: function-use-after-free
```

File: c/lang/security/function-use-after-free.c [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+int bad_code2() {
+    NAME *var;
+    char buf[10];
+    var = (NAME *)malloc(sizeof(struct name));
+    free(var);
+    // ruleid: use-after-free
+    strcpy(buf, (char*)*var);
+    // ruleid: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:36]:
```suggestion
    // ruleid: function-use-after-free
```

File: c/lang/security/function-use-after-free.c [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    other_func((char*)var);
+    return 0;
+}
+
+int bad_code2() {
+    NAME *var;
+    char buf[10];
+    var = (NAME *)malloc(sizeof(struct name));
+    free(var);
+    // ruleid: use-after-free
```
Comment (kurt-r2c) [inline c/lang/security/function-use-after-free.c:34]:
```suggestion
    // ruleid: function-use-after-free
```

**Review by kurt-r2c — CHANGES_REQUESTED** [review]

**Pull request closed without merging**